tools/lkl/lib/hijack/init.c

```diff
@@ -139,65 +139,52 @@ static void mount_cmds_exec(char *_cmds, int (*callback)(char*))
 	free(cmds);
 }
 
-int init_tap(char *tap, char *mac_str)
-{
-	struct ifreq ifr = {
-		.ifr_flags = IFF_TAP | IFF_NO_PI,
-	};
-	union lkl_netdev nd;
-	__lkl__u8 mac[LKL_ETH_ALEN] = {0};
-	int ret = -1;
-
-	strncpy(ifr.ifr_name, tap, IFNAMSIZ);
-
-	nd.fd = open("/dev/net/tun", O_RDWR|O_NONBLOCK);
-	if (nd.fd < 0) {
-		fprintf(stderr, "failed to open tap: %s\n", strerror(errno));
-		return -1;
-	}
-
-	ret = ioctl(nd.fd, TUNSETIFF, &ifr);
-	if (ret < 0) {
-		fprintf(stderr, "failed to attach to %s: %s\n",
-			ifr.ifr_name, strerror(errno));
-		return -1;
-	}
-
-	ret = parse_mac_str(mac_str, mac);
-
-	if (ret < 0) {
-		fprintf(stderr, "failed to parse mac\n");
-		return -1;
-	} else if (ret > 0) {
-		ret = lkl_netdev_add(nd, mac);
-	} else {
-		ret = lkl_netdev_add(nd, NULL);
-	}
-
-	if (ret < 0) {
-		fprintf(stderr, "failed to add netdev: %s\n",
-			lkl_strerror(ret));
-		return -1;
-	}
-
-	return ret;
-}
-
 void __attribute__((constructor(102)))
 hijack_init(void)
 {
 	int ret, i, dev_null, nd_id = -1, nd_ifindex = -1;
+	/* OBSOLETE: should use IFTYPE and IFPARAMS */
 	char *tap = getenv("LKL_HIJACK_NET_TAP");
+	char *iftype = getenv("LKL_HIJACK_NET_IFTYPE");
+	char *ifparams = getenv("LKL_HIJACK_NET_IFPARAMS");
 	char *mtu_str = getenv("LKL_HIJACK_NET_MTU");
+	__lkl__u8 mac[LKL_ETH_ALEN] = {0};
 	char *ip = getenv("LKL_HIJACK_NET_IP");
 	char *mac_str = getenv("LKL_HIJACK_NET_MAC");
 	char *netmask_len = getenv("LKL_HIJACK_NET_NETMASK_LEN");
 	char *gateway = getenv("LKL_HIJACK_NET_GATEWAY");
 	char *debug = getenv("LKL_HIJACK_DEBUG");
 	char *mount = getenv("LKL_HIJACK_MOUNT");
+	struct lkl_netdev *nd = NULL;
 
 	if (tap) {
-		nd_id = init_tap(tap, mac_str);
+		fprintf(stderr,
+			"WARN: variable LKL_HIJACK_NET_TAP is now obsoleted.\n"
+			"      please use LKL_HIJACK_NET_IFTYPE and "
+			"LKL_HIJACK_NET_IFPARAMS instead.\n");
+		nd = lkl_netdev_tap_create(ifparams);
+	}
+
+	if (!tap && iftype && ifparams && (strncmp(iftype, "tap", 3) == 0))
```

Comment:
`if (!nd` might be better.

Also, by using `strncmp`, do we really want to tolerate `iftype`s such as `tapXXX`?